File: 414_ThirdMaximumNumber/414_ThirdMaximumNumber.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var thirdMax = function (nums) {
  let first = (second = third = -Infinity);
  let compare = new Set(nums);
  for (let num of compare) {
    if (num > third) {
      first = second;
      second = third;
      third = num;
    } else if (num > second) {
      first = second;
      second = num;
    } else if (num > first) {
      first = num;
    }
  }

  return first != -Infinity ? first : third;
};

File: 414_ThirdMaximumNumber/414_ThirdMaximumNumber.py
class Solution:
       def thirdMax(self, nums: List[int]) -> int:
        first = second = third = -math.inf
        for num in nums:
            if num > third and num != second and num != first:
                first, second, third = second, third, num
            elif num > second and num != third and num != first:
                first, second = second, num
            elif num > first and num != second and num != third:
                first = num
        return first if first != -math.inf else third
       
       """
       
       getting this done in o(n)
       """